presto-main/src/main/java/com/facebook/presto/operator/aggregation/noisyaggregation/sketch/SfmSketch.java

```diff
@@ -0,0 +1,426 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator.aggregation.noisyaggregation.sketch;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import io.airlift.slice.BasicSliceInput;
+import io.airlift.slice.DynamicSliceOutput;
+import io.airlift.slice.Murmur3Hash128;
+import io.airlift.slice.SizeOf;
+import io.airlift.slice.Slice;
+import org.openjdk.jol.info.ClassLayout;
+
+import javax.annotation.concurrent.NotThreadSafe;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+/**
+ * SfmSketch is a sketch for distinct counting, very similar to HyperLogLog.
+ * This sketch is introduced as the Sketch-Flip-Merge (SFM) summary in the paper
+ * <a href="https://arxiv.org/pdf/2302.02056.pdf">Sketch-Flip-Merge: Mergeable Sketches for Private Distinct Counting</a>.
+ * <p>
+ * The primary differences between SfmSketch and HyperLogLog are that
+ * (a) SfmSketch supports differential privacy, and
+ * (b) where HyperLogLog tracks only max observed bucket values, SfmSketch tracks all bucket values observed.
+ * <p>
+ * This means that SfmSketch is a larger sketch than HyperLogLog, but offers the ability to store completely
+ * DP sketches with a fixed, public hash function while maintaining accurate cardinality estimates.
+ * <p>
+ * SfmSketch is created in a non-private mode. Privacy must be enabled through the enablePrivacy() function.
+ * Once made private, the sketch becomes immutable. Privacy is quantified by the parameter epsilon.
+ * <p>
+ * When epsilon > 0, the sketch is epsilon-DP, and bits are randomized to preserve privacy.
+ * When epsilon == NON_PRIVATE_EPSILON, the sketch is not private, and bits are set deterministically.
+ * <p>
+ * The best accuracy comes with NON_PRIVATE_EPSILON. For private epsilons, larger gives more accuracy,
+ * while smaller gives more privacy.
+ */
+@NotThreadSafe
+public class SfmSketch
+{
+    public static final double NON_PRIVATE_EPSILON = Double.POSITIVE_INFINITY;
+
+    private static final byte FORMAT_TAG = 7;
+    private static final int MAX_ESTIMATION_ITERATIONS = 1000;
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(SfmSketch.class).instanceSize();
+
+    private final int indexBitLength;
+    private final int precision;
+    private final RandomizationStrategy randomizationStrategy;
+
+    private double randomizedResponseProbability;
```

Comment:
How difficuilt would it be to refactor the tests and make these two fields final?

Reply:
Making `randomizedResponseProbability` final would require returning copies rather than mutating in `enablePrivacy` and `mergeWith`. `mergeWith` in particular was designed to mutate to match the functionality in `HyperLogLog`, `QuantileDigest`, etc.

`bitmap` seems like it should be final, but I'll take a closer look.

Reply:
`bitmap` is now final.